pilot/pkg/config/kube/gateway/gateway_collection.go

```diff
@@ -50,8 +53,168 @@ func (g Gateway) Equals(other Gateway) bool {
+		g.GatewayParent == other.GatewayParent &&
+		g.Valid == other.Valid // TODO: ok to ignore parent/parentInfo?
+}
+
+func ListenerSetCollection(
+	listenerSets krt.Collection[*gatewayx.XListenerSet],
+	gateways krt.Collection[*gateway.Gateway],
+	gatewayClasses krt.Collection[GatewayClass],
+	namespaces krt.Collection[*corev1.Namespace],
+	grants ReferenceGrants,
+	secrets krt.Collection[*corev1.Secret],
+	domainSuffix string,
+	gatewayContext krt.RecomputeProtected[*atomic.Pointer[GatewayContext]],
+	tagWatcher krt.RecomputeProtected[revisions.TagWatcher],
+	opts krt.OptionsBuilder,
+) (
+	krt.StatusCollection[*gatewayx.XListenerSet, gatewayx.ListenerSetStatus],
+	krt.Collection[ListenerSet],
+) {
+	statusCol, gw := krt.NewStatusManyCollection(listenerSets,
+		func(ctx krt.HandlerContext, obj *gatewayx.XListenerSet) (*gatewayx.ListenerSetStatus, []ListenerSet) {
+			// We currently depend on service discovery information not know to krt; mark we depend on it.
+			context := gatewayContext.Get(ctx).Load()
+			if context == nil {
+				return nil, nil
+			}
+			if !tagWatcher.Get(ctx).IsMine(obj.ObjectMeta) {
+				return nil, nil
+			}
+			result := []ListenerSet{}
+			ls := obj.Spec
+			status := obj.Status.DeepCopy()
+
+			p := ls.ParentRef
+			if normalizeReference(p.Group, p.Kind, gvk.KubernetesGateway) != gvk.KubernetesGateway {
+				// Cannot report status since we don't know if it is for us
+				return nil, nil
+			}
+
+			pns := ptr.OrDefault(p.Namespace, gatewayx.Namespace(obj.Namespace))
+			parentGwObj := ptr.Flatten(krt.FetchOne(ctx, gateways, krt.FilterKey(string(pns)+"/"+string(p.Name))))
+			if parentGwObj == nil {
+				// Cannot report status since we don't know if it is for us
+				return nil, nil
+			}
+
+			class := fetchClass(ctx, gatewayClasses, parentGwObj.Spec.GatewayClassName)
+			if class == nil {
+				// Cannot report status since we don't know if it is for us
+				return nil, nil
+			}
+
+			controllerName := class.Controller
+			classInfo, f := classInfos[controllerName]
+			if !f {
+				// Cannot report status since we don't know if it is for us
+				return nil, nil
+			}
+			if !classInfo.supportsListenerSet {
+				reportUnsupportedListenerSet(class.Name, status, obj)
+				return status, nil
+			}
+
+			if !namespaceAcceptedByAllowListeners(obj.Namespace, parentGwObj, func(s string) *corev1.Namespace {
+				return ptr.Flatten(krt.FetchOne(ctx, namespaces, krt.FilterKey(s)))
+			}) {
+				reportNotAllowedListenerSet(status, obj)
+				return status, nil
+			}
+
+			gatewayServices, err := extractGatewayServices(domainSuffix, parentGwObj, classInfo)
+			if len(gatewayServices) == 0 && err != nil {
+				// Short circuit if it's a hard failure
+				reportListenerSetStatus(context, parentGwObj, obj, status, gatewayServices, nil, err)
+				return status, nil
+			}
+
+			servers := []*istio.Server{}
+			for i, l := range ls.Listeners {
+				port, err := detectListenerPort(l)
+				_ = err // TODO
```

Comment:
?

Reply:
Fixed, but cannot be tested until https://github.com/kubernetes-sigs/gateway-api/pull/3750 lands (it is impossible to get this error in 1.3 but will be possible in future